mpisppy/utils/stoch_admmWrapper.py

```diff
@@ -0,0 +1,231 @@
+            depth = len(s._mpisppy_node_list)+1
+            varlist = [[] for _ in range(depth)]
+            if s._mpisppy_probability == "uniform": #if there is a two-stage scenario defined by sputils.attach_root_node
+                s._mpisppy_probability = 1/len(self.stoch_scenario_names)
+
+            self.varprob_dict[s] = list()
+            for vstr in all_consensus_vars.keys():
+                stage = all_consensus_vars[vstr]
+                v = s.find_component(vstr)
+                var_stage_tuple = vstr, stage
+                if var_stage_tuple in self.consensus_vars[admm_subproblem_name]:
+                    if v is not None:
+                        # variables that should be on the model
+                        if stage == depth: 
+                            # The node is a stochastic scenario, the probability at this node has not yet been defined 
+                            cond_prob = 1.
+                        else:
+                            prob_node = np.prod([s._mpisppy_node_list[ancestor_stage-1].cond_prob for ancestor_stage in range(1,stage+1)])
+                            # conditional probability of the scenario at the node (without considering the leaves as probabilities)
+                            cond_prob = s._mpisppy_probability/prob_node 
+                        self.varprob_dict[s].append((id(v),cond_prob/(self.consensus_vars_number[vstr])))
+                        # s._mpisppy_probability has not yet been divided by the number of subproblems, it the probability of the
+                        # stochastic scenario
+                    else:
+                        error_list1.append((sname,vstr))
+                else:
+                    if v is None:
+                        # This var will not be indexed but that might not matter??
+                        # Going to replace the brackets
+                        v2str = vstr.replace("[","__").replace("]","__") # To distinguish the consensus_vars fixed at 0
+                        v = pyo.Var()
+                        
+                        ### Lines equivalent to setattr(s, v2str, v) without warning
+                        #s.del_component(v2str) UNUSEFUL
+                        s.add_component(v2str, v) 
+                        #is the consensus variable should be earlier, then its not added in the good place, or is it?
+
+                        v.fix(0)
+                        self.varprob_dict[s].append((id(v),0))
+                    else:
+                        error_list2.append((sname,vstr))
+                varlist[stage-1].append(v)
+
+            # Create the new scenario tree node for admm_consensus
+            assert hasattr(s,"_mpisppy_node_list"), f"the scenario {sname} doesn't have any _mpisppy_node_list attribute"
+            parent = s._mpisppy_node_list[-1]
+            objfunc = pyo.Expression(expr=0) # The cost is spread on the branches which are the subproblems
```

Comment:
Not sure where objfunc is used below - it is only passed in as an argument below.